### PR TITLE
feat: upgrade to angular 13

### PR DESCRIPTION
**Pull request by pavankjadda**

### CODE REVIEW

**Comment by pavankjadda** [comment]
@varnastadeus when you get a chance review this and approve the PR

**Comment by inexuscore** [comment]
Any idea when Angular 13 support will be available? Kinda desperately need to update to Angular 13 ASAP. Thank you.

**Comment by pavankjadda** [comment]
> Any idea when Angular 13 support will be available? Kinda desperately need to update to Angular 13 ASAP. Thank you.

Hopefully soon. @varnastadeus reviewing the code now.

**Comment by inexuscore** [comment]
@pavankjadda That's awesome, thank you!

File: angular.json
Comment (terencehonles) [inline angular.json:36]: Should eslint not be enabled instead of just dropping tslint? Specifically builder: `@angular-eslint/builder:lint`

Comment (terencehonles) [reply angular.json:36]: This looks like why the CI is failing too since there is now no lint command.

Comment (varnastadeus) [reply angular.json:36]: @pavankjadda can you address this?

Comment (pavankjadda) [reply angular.json:36]: shall I add `@angular-eslint` to the project and remove tslint?

Comment (pavankjadda) [reply angular.json:36]: okay. I added `angular-eslint` to the project

File: src/ng-option-highlight/.eslintrc.json
Comment (terencehonles) [inline src/ng-option-highlight/.eslintrc.json:67]: Should we fix these so we don't have to turn these off? I assume they must be triggering errors if they're not matching what was in the previous [tslist.json](https://github.com/ng-select/ng-select/pull/1996/files#diff-b09e073e9fc2668f089c105170728e74f159eccce9163ec6f979b6d182916d70)

File: src/ng-select/.eslintrc.json
Comment (terencehonles) [inline src/ng-select/.eslintrc.json:35]: do you need to specify these since they are defined in [.eslintrc.json](https://github.com/ng-select/ng-select/pull/1996/files#diff-6884918dc8291219be508e05e28965b958c734def85324f3b53858ea4702090f) and it looks like this should extend that config? (The same applies to [src/ng-option-highlight/.eslintrc.json](https://github.com/ng-select/ng-select/pull/1996/files#diff-9aebdea382bb6ae71b2be933869895a6bc82eb729f67492cf024c90dc118d860))

File: .eslintrc.json
Comment (terencehonles) [inline .eslintrc.json:19]: Does recommended cover all the rules in [tslint.json](https://github.com/ng-select/ng-select/pull/1996/files#diff-9db81b3f77968b3ad243c25f280c4e80dd20b2e5770bf5606ccdb4ecb8af9ce8)?

**Review by terencehonles — COMMENTED** [review]
@pavankjadda thanks for adding the lint config back, a few questions on it :slightly_smiling_face:

File: src/ng-option-highlight/.eslintrc.json
Comment (pavankjadda) [reply src/ng-option-highlight/.eslintrc.json:67]: `@angular-eslint` library converted all rules from `tslint.json` to this file. But to be honest I haven't looked at the file closely, since they do good job of it. But I am not sure if we need to tackle this in this PR or have separate for it.

File: .eslintrc.json
Comment (pavankjadda) [reply .eslintrc.json:19]: I ran lint based on this config, I got following report

```
/Users/jaddap2/WebstormProjects/ng-select/src/ng-select/lib/ng-dropdown-panel.component.ts
  35:15  error  The selector should start with one of these prefixes: "lib" (https://angular.io/guide/styleguide#style-02-07)  @angular-eslint/component-selector

/Users/jaddap2/WebstormProjects/ng-select/src/ng-select/lib/ng-option.component.ts
  14:15  error  The selector should start with one of these prefixes: "lib" (https://angular.io/guide/styleguide#style-02-07)  @angular-eslint/component-selector

/Users/jaddap2/WebstormProjects/ng-select/src/ng-select/lib/ng-select.component.spec.ts
  4150:7  error  Component class names should end with one of these suffixes: "Component" (https://angular.io/styleguide#style-02-03)  @angular-eslint/component-class-suffix
  4241:7  error  Component class names should end with one of these suffixes: "Component" (https://angular.io/styleguide#style-02-03)  @angular-eslint/component-class-suffix
  4248:7  error  Component class names should end with one of these suffixes: "Component" (https://angular.io/styleguide#style-02-03)  @angular-eslint/component-class-suffix

/Users/jaddap2/WebstormProjects/ng-select/src/ng-select/lib/ng-select.component.ts
   62:15  error    The selector should start with one of these prefixes: "lib" (https://angular.io/guide/styleguide#style-02-07)           @angular-eslint/component-selector
   72:5   error    Use @HostBinding or @HostListener rather than the `host` metadata property (https://angular.io/styleguide#style-06-03)  @angular-eslint/no-host-metadata-property
  258:5   warning  Lifecycle interface 'OnInit' should be implemented for method 'ngOnInit'. (https://angular.io/styleguide#style-09-01)   @angular-eslint/use-lifecycle-interface

/Users/jaddap2/WebstormProjects/ng-select/src/ng-select/lib/ng-templates.directive.ts
   4:24  error  The selector should start with one of these prefixes: "lib" (https://angular.io/guide/styleguide#style-02-08)  @angular-eslint/directive-selector
  18:24  error  The selector should be camelCase (https://angular.io/guide/styleguide#style-02-06)                             @angular-eslint/directive-selector
  23:24  error  The selector should be camelCase (https://angular.io/guide/styleguide#style-02-06)                             @angular-eslint/directive-selector
  28:24  error  The selector should be camelCase (https://angular.io/guide/styleguide#style-02-06)                             @angular-eslint/directive-selector
  33:24  error  The selector should be camelCase (https://angular.io/guide/styleguide#style-02-06)                             @angular-eslint/directive-selector
  38:24  error  The selector should be camelCase (https://angular.io/guide/styleguide#style-02-06)                             @angular-eslint/directive-selector
  43:24  error  The selector should be camelCase (https://angular.io/guide/styleguide#style-02-06)                             @angular-eslint/directive-selector
  48:24  error  The selector should be camelCase (https://angular.io/guide/styleguide#style-02-06)                             @angular-eslint/directive-selector
  53:24  error  The selector should be camelCase (https://angular.io/guide/styleguide#style-02-06)                             @angular-eslint/directive-selector
  58:24  error  The selector should be camelCase (https://angular.io/guide/styleguide#style-02-06)                             @angular-eslint/directive-selector
  63:24  error  The selector should be camelCase (https://angular.io/guide/styleguide#style-02-06)                             @angular-eslint/directive-selector
  68:24  error  The selector should be camelCase (https://angular.io/guide/styleguide#style-02-06)                             @angular-eslint/directive-selector


```
we can the following errors since they are false positive

- The selector should be camelCase
-  The selector should start with one of these prefixes: "lib"

but we may want to resolve this
```
   72:5   error    Use @HostBinding or @HostListener rather than the `host` metadata property (https://angular.io/styleguide#style-06-03)  @angular-eslint/no-host-metadata-property
```

**Comment by terencehonles** [comment]
@pavankjadda I created https://github.com/pavankjadda/ng-select/pull/1 against your branch that this PR is created off of including my suggestions and additional eslint config changes. You can merge the PR to update this PR with my suggestions.

File: .eslintrc.json
Comment (terencehonles) [reply .eslintrc.json:19]: Fixed on https://github.com/pavankjadda/ng-select/pull/1

**Comment by pavankjadda** [comment]
@terencehonles merged your changes. All files passing lint rules. This is good to go now.

**Comment by pavankjadda** [comment]
@varnastadeus please review this

**Comment by terencehonles** [comment]
> @terencehonles merged your changes. All files passing lint rules. This is good to go now.

Thanks for the follow up @pavankjadda. @varnastadeus I followed up to confirm the lint rules match the original intent (some things slipped through because tslint didn't catch everything it was configured to)

File: angular.json
Comment (terencehonles) [inline angular.json:3]: This is probably fine, but it probably can also be dropped. Not sure if you have a preference @varnastadeus

**Review by terencehonles — APPROVED** [review]
Finishing a full review, LGTM!

**Comment by pavankjadda** [comment]
> Finishing a full review, LGTM!

Do you have access to merge this and release new version?

**Comment by terencehonles** [comment]
> > Finishing a full review, LGTM!
> 
> Do you have access to merge this and release new version?

Nope, that would be @varnastadeus, just letting them know I thought the PR looked good :slightly_smiling_face:

**Comment by pavankjadda** [comment]
@varnastadeus When you get a chance please review and approve this.

**Comment by varnastadeus** [comment]
> @varnastadeus When you get a chance please review and approve this.

@pavankjadda tests are not running in the CI job. Can you check on this?

**Comment by pavankjadda** [comment]
> > @varnastadeus When you get a chance please review and approve this.
> 
> 
> 
> @pavankjadda tests are not running in the CI job. Can you check on this?

I am able to run them on local machine. It seems some kind of issue with GitHub actions when using Karma

**Comment by varnastadeus** [comment]
> > > @varnastadeus When you get a chance please review and approve this.
> > 
> > 
> > @pavankjadda tests are not running in the CI job. Can you check on this?
> 
> I am able to run them on local machine. It seems some kind of issue with GitHub actions when using Karma

@pavankjadda I am having the same error while running `yarn test` locally.

**Comment by pavankjadda** [comment]
> > > > @varnastadeus When you get a chance please review and approve this.
> 
> > > 
> 
> > > 
> 
> > > @pavankjadda tests are not running in the CI job. Can you check on this?
> 
> > 
> 
> > I am able to run them on local machine. It seems some kind of issue with GitHub actions when using Karma
> 
> 
> 
> @pavankjadda I am having the same error while running `yarn test` locally.

Interesting. Let me check

**Comment by pavankjadda** [comment]
@varnastadeus would you mind enabling CI workflow on each commit and pull request. I just pushed an update and I don't know if it passes CI

**Comment by pavankjadda** [comment]
I see all tests passing (272/273) expect this one 
```
            it('should allow edit search if option selected & input focused', fakeAsync(() => {
                const fixture = createTestingModule(
                    NgSelectTestComponent,
                    `<ng-select [items]="cities"
                        [typeahead]="filter"
                        [editableSearchTerm]="true"
                        bindValue="id"
                        bindLabel="name"
                        [(ngModel)]="selectedCity">
                    </ng-select>`);
                expect(fixture.componentInstance.select.editableSearchTerm).toBeTruthy();
                const select = fixture.componentInstance.select;
                const input = select.searchInput.nativeElement;
                const selectedCity = fixture.componentInstance.cities[0];
                fixture.componentInstance.selectedCity = selectedCity.id;
                tickAndDetectChanges(fixture);
                input.focus();
                tickAndDetectChanges(fixture);
                expect(select.searchTerm).toEqual(selectedCity.name);
                expect(input.value).toEqual(selectedCity.name);
            }));
```

I see the following error message in Console. I wonder if this was introduced by changes from @terencehonles. But I am not sure. 

```
	Error: Expected null to equal 'Vilnius'.
	    at <Jasmine>
	    at UserContext.<anonymous> (src/ng-select/lib/ng-select.component.spec.ts:3054:43)
	    at UserContext.fakeAsyncFn (node_modules/zone.js/fesm2015/zone-testing.js:1968:1)
	    at ZoneDelegate.invoke (node_modules/zone.js/fesm2015/zone.js:372:1)
	Error: Expected '' to equal 'Vilnius'.
	    at <Jasmine>
	    at UserContext.<anonymous> (src/ng-select/lib/ng-select.component.spec.ts:3055:37)
	    at UserContext.fakeAsyncFn (node_modules/zone.js/fesm2015/zone-testing.js:1968:1)
	    at ZoneDelegate.invoke (node_modules/zone.js/fesm2015/zone.js:372:1)


```

File: src/ng-select/lib/ng-select.component.spec.ts
Comment (terencehonles) [inline src/ng-select/lib/ng-select.component.spec.ts:3038]: I was able to fix https://github.com/ng-select/ng-select/pull/1996#issuecomment-969047202

by adding a `tick(200)` before line 3054 (which I can't comment on in this diff). I'm not sure why the compontent is taking more time to resolve though :thinking:

File: src/ng-option-highlight/karma.conf.js
Comment (terencehonles) [inline src/ng-option-highlight/karma.conf.js:13]: I think you can remove this based on adding `karma-coverage` and https://github.com/angular/angular-cli/blob/master/CHANGELOG.md#angular-devkitbuild-angular-3

```suggestion
```

File: src/ng-select/karma.conf.js
Comment (terencehonles) [inline src/ng-select/karma.conf.js:13]: I think you can remove this based on adding `karma-coverage` and https://github.com/angular/angular-cli/blob/master/CHANGELOG.md#angular-devkitbuild-angular-3

```suggestion
```

**Review by terencehonles — COMMENTED** [review]
I believe you can probably remove from the karma config and the `package.json` `karma-coverage-istanbul-reporter`

File: src/ng-select/karma.conf.js
Comment (pavankjadda) [reply src/ng-select/karma.conf.js:13]: agree

File: src/ng-select/lib/ng-select.component.spec.ts
Comment (pavankjadda) [reply src/ng-select/lib/ng-select.component.spec.ts:3038]: Interesting. For me I had to add `tick(2000)` not `tick(200)`. Don't why.

**Comment by pavankjadda** [comment]
@varnastadeus all the tests passing on my local machine. You can review this now.

File: package.json
Comment (terencehonles) [inline package.json:81]: @pavankjadda you mind running `yarn remove karma-coverage-istanbul-reporter`?

Comment (pavankjadda) [reply package.json]: Done

File: src/ng-select/lib/ng-select.component.spec.ts
Comment (pavankjadda) [reply src/ng-select/lib/ng-select.component.spec.ts:3038]: > I was able to fix [#1996 (comment)](https://github.com/ng-select/ng-select/pull/1996#issuecomment-969047202)
> 
> by adding a `tick(200)` before line 3054 (which I can't comment on in this diff). I'm not sure why the compontent is taking more time to resolve though 🤔

This test seems to be flakey. Adding `tick(2000)` sometimes passes and fails other times.

**Comment by varnastadeus** [comment]
fix coverage reporter

**Comment by varnastadeus** [comment]
```diff
diff --git a/src/ng-select/karma.conf.js b/src/ng-select/karma.conf.js
--- src/ng-select/karma.conf.js
+++ src/ng-select/karma.conf.js
@@ -14,11 +14,11 @@
         ],
         client: {
             clearContext: false // leave Jasmine Spec Runner output visible in browser
         },
-        coverageIstanbulReporter: {
+        coverageReporter: {
             dir: require('path').join(__dirname, '../../coverage/ng-select'),
-            reports: ['lcovonly'],
+            reporters: ['lcovonly'],
             fixWebpackSourcePaths: true
         },
         reporters: ['progress', 'kjhtml'],
         port: 9876,

```

**Review by varnastadeus — APPROVED** [review]

**Comment by terencehonles** [comment]
@varnastadeus I was about to do the same :sweat_smile: it looks like we may want to also pull in https://github.com/angular/angular-cli/commit/ef8815d0434836f2d8119e91a7bc09742ff77d37 but when updating to Angular 13.0.2 I noticed that I had to use `--force` since `@ng-bootstrap/ng-bootstrap` isn't already updated for Angular 13 https://github.com/ng-bootstrap/ng-bootstrap/issues/4178 we may want to wait for that before merging this (or at least packaging this).

**Comment by terencehonles** [comment]
These are the changes I made https://github.com/pavankjadda/ng-select/pull/2

**Comment by terencehonles** [comment]
Since you have push access to @pavankjadda's branch you can fast forward it with `git fetch git@github.com:terencehonles/ng-select.git update-karma-angular-and-ng-bootstrap && git push git@github.com:pavankjadda/ng-select.git FETCH_HEAD:angular13-support` I can't do that however since I'm not a maintainer of this repo and don't inherit that permission.

**Comment by pavankjadda** [comment]
> ```diff
> diff --git a/src/ng-select/karma.conf.js b/src/ng-select/karma.conf.js
> --- src/ng-select/karma.conf.js
> +++ src/ng-select/karma.conf.js
> @@ -14,11 +14,11 @@
>          ],
>          client: {
>              clearContext: false // leave Jasmine Spec Runner output visible in browser
>          },
> -        coverageIstanbulReporter: {
> +        coverageReporter: {
>              dir: require('path').join(__dirname, '../../coverage/ng-select'),
> -            reports: ['lcovonly'],
> +            reporters: ['lcovonly'],
>              fixWebpackSourcePaths: true
>          },
>          reporters: ['progress', 'kjhtml'],
>          port: 9876,
> ```

Thanks for this. I still see CI failing in `Coverall` stage. Any idea on how to fix it?

**Comment by pavankjadda** [comment]
@varnastadeus never mind. I updated karma config based on latest [configuration changes](https://github.com/karma-runner/karma-coverage/blob/master/docs/configuration.md#subdir). I tested them in another branch in my fork. All works fine and CI is passing. You can merge the PR now

**Review by varnastadeus — APPROVED** [review]

**Comment by varnastadeus** [comment]
thanks again for the PR guys 🎉

**Comment by pavankjadda** [comment]
> thanks again for the PR guys 🎉

No problem. Happy to contribute. I [see](https://github.com/ng-select/ng-select/runs/4227323405?check_suite_focus=true#step:5:18) GitHub actions failed to release new version. You can see the [build log](https://github.com/ng-select/ng-select/runs/4227323405?check_suite_focus=true#step:5:18) here

**Comment by varnastadeus** [comment]
@pavankjadda I have addressed those issues here #2004

**Pull request merged**
